numba/npyufunc/dufunc.py

```diff
@@ -0,0 +1,33 @@
+from __future__ import absolute_import, print_function, division
+import numpy
+
+from numba import jit, typeof, utils
+from .. import numpy_support
+from . import _internal, ufuncbuilder
+
+class DUFunc(_internal._DUFunc):
+    def __init__(self, py_func, **kws):
+        dispatcher = jit(target='npyufunc')(py_func)
+        super(DUFunc, self).__init__(dispatcher, **kws)
+        # Clean up _DUFunc keyword arguments, and assume anything that
+        # remains is a compiler option.
+        kws.pop('identity', None)
+        kws.pop('nin', None)
+        kws.pop('nout', None)
+        self.targetoptions = kws
+
+    def _compile_for_args(self, *args, **kws):
+        assert len(kws) == 0
+        argtys = tuple(numpy_support.map_arrayscalar_type(arg)
+                       if numpy_support.is_arrayscalar(arg)
+                       else typeof(arg)
+                       for arg in args[:self.ufunc.nin])
+        if all(hasattr(argty, 'dtype') for argty in argtys):
```

Comment:
This does look a bit weird, but that's because we have to determine if we are using a Numpy scalar or not.  We have to use Numpy's typing judgments on scalars because Numba makes different choices on how to type values.  Regarding how I expressed this: I thought list comprehensions were idiomatic at this point.